#+AUTHOR: Jan Christian Kaessens <j.kaessens@ikmb.uni-kiel.de>
#+TITLE: GWAS QC Pipeline
#+STARTUP: showall

/Note: This is currently under heavy development. Is likely that this documentation does not match the actual source code. It will soon, though./

This pipeline is divided into several sub-pipelines that do quality control on
GWAS data sets. It can be run on any platform that supports Singularity, i.e.
the =rzcluster=, a local workstation, or the experimental IKMB OpenStack
elasticluster.

* Setup
** Container Base Image Bootstrapping
  Before the container can be created, a base image must be built with the
  initial operating system. For this, you will need a computer with root access and Singularity 2.4 installed.
#+BEGIN_QUOTE
=singularity build singularity/qc.img singularity/QC.def
#+END_QUOTE
** Container Bootstrapping
  Before you can run any of the pipelines, you will need to bootstrap the
  pipeline container on a computer you have root access on. Perform the
  following commands to create the container =qc.img=:
#+BEGIN_QUOTE
=singularity build singularity/baseimage.img singularity/baseimage.def=
=singularity build singularity/qc.img singularity/QC.def=
#+END_QUOTE
  Now move the container to the place where you want to run the pipelines, i.e.
the =rzcluster= head node or the elasticluster frontend node and you're ready to go.

If you want to install additional software, change QC.def and re-build only the
=qc.img= image. You should not normally need to update the base image, but if
you do, you will also need to re-build the =qc.img= image.

** Configuration Adjustment
   You should also take a look at the various configuration files and adjust some file paths:
*** =nextflow.config=
    Choose an environment configuration. You will most likely want the
    =rzcluster-singularity.config=, so be sure that this is the only one
    uncommented.
*** =config/singularity.config=
    In the singularity stanza, change the =runOptions= key in the following manner:
#+BEGIN_QUOTE
=runOptions = "--bind /home/myusername -H /path/to/pipeline-root/:/home/sukmb388/"=
#+END_QUOTE
    Be sure to set =/home/myusername= to your actual home folder and
    =/path/to/pipeline-root= to the folder where you put the pipeline stage
    scripts. The current configuration expects this folder to contain the
    =PoAtools_2= and =immunochip_batches= folders from David Ellinghaus'
    =PoAtools_2= package.
*** =config/{QC-Rs,QCI,SampleQCI}.config=
    These files have been auto-generated from the =PoAtools_2= configuration. If
    you need changes to the workflow parameters, thresholds, etc., this is the
    place to go.

  You will probably notice, that most of the config keys contain the
  =/home/sukmb388= path specification. This is okay even if you are not
  =sukmb388=, because singularity's =runOptions= make the pipeline scripts
  believe you are. There is no need to change that. In fact, you would probably break things.

* Running

In some environments (i.e. =rzcluster=), you will need to load some modules: =IKMB=, =Nextflow= and =singularity2.3=.

Actually running the pipeline script is rather straight-forward:
#+BEGIN_QUOTE
~nextflow -c config/QC-Rs.config -c config/set_cvidv2_rs.config run QC-Rs.nf --input=./CVID_Ichip-orig_DvH-Opti_v2/CVID_Ichip-orig_DvH-Opti_v2 --output=out-RS~
#+END_QUOTE

You only need to specifiy the stage configuration =QC-Rs.config=, the data
set-specific configuration =set_cvidv2_rs.config=, the actual script =QC-RS.nf=,
the input files and output folder.

Unfortunately, not much error checking is currently done and if errors appear,
they will likely be somewhat cryptic. This will change in the future.


* Development and Implementation Status
  | Original file                 | New file       | Implemented     | Verification (CVID) | Verification (GSA)   |
  |-------------------------------+----------------+-----------------+---------------------+----------------------|
  | =Rs.py=                       | =QC-Rs.nf=     | <2017-05-15 Mo> |                     | <2017-11-08 Mi> pass |
  | =SNPQCI_parallel_part1.py=    | =QCI.nf=       | <2017-05-17 Mi> |                     | <2017-11-09 Do> pass |
  | =SNPQCI_parallel_part2.py=    | =QCI.nf=       | <2017-06-28 Mi> |                     | <2017-11-09 Do> pass |
  | =SampleQCI_parallel_part1.py= | =SampleQCI.nf= | <2017-08-02 Mi> |                     |                      |
  | =SampleQCI_parallel_part2.py= | =SampleQCI.nf= |                 |                     |                      |
  | =SampleQCI_parallel_part3.py= | =SampleQCI.nf= |                 |                     |                      |
  | =SampleQCI_parallel_part4.py= | =SampleQCI.nf= |                 |                     |                      |

** QC-Rs.nf
  - [X] =generate_annotations=
  - [X] =generate_flipfile=
  - [X] =find_duplicates=
  - [X] =find_nn=
  - [X] =merge_exclude_list=
  - [X] =plink_exclude=
** QCI.nf
   | Process                      | Verified | Main result file                         | Hash                             | V. Result file                                 | V. Hash                          | Problem             |
   |------------------------------+----------+------------------------------------------+----------------------------------+------------------------------------------------+----------------------------------+---------------------|
   | merge_batches                | ok       | DKTNF_POPGEN_GS_QCI_merged.bed           | d5823498ff41ac4bc06e4587b5741138 | SNPQCI/DKTNF_POPGEN_GS_unQCed.bed              | d5823498ff41ac4bc06e4587b5741138 |                     |
   | det_miss_per_batch           | ok       | cluster_file                             | 1dc83dba06cc0d8fd8e672e5f88aaca1 | DKTNF_POPGEN_GS_unQCed.cluster-file.txt        | 1dc83dba06cc0d8fd8e672e5f88aaca1 |                     |
   |                              | ok       | missingness_per_batch.lmiss              |                                  | DKTNF_POPGEN_GS_unQCed.per_batch.lmiss         | 6c4e7378098a6f25e318d8b13ede18df | Plink 1.7/1.9       |
   |                              | ok       | missingness_per_batch.imiss              |                                  | DKTNF_POPGEN_GS_unQCed.per_batch.imiss         | c1e86f93dd40a510e8bd6e59a9e9fb95 | Plink 1.7/1.9       |
   |                              | ok       | missingness-excludes-perbatch            | 73c5d0d9add885b3c2bb6e538016f7d3 | .lmiss.Variant-exclude-list.SNPQCI.4.txt       | 73c5d0d9add885b3c2bb6e538016f7d3 | Python par Ãœbergabe |
   | det_miss_entire              | ok       | missingness_entire.imiss                 | c1e86f93dd40a510e8bd6e59a9e9fb95 | DKTNF_POPGEN_GS_unQCed.entire_collection.imiss | c1e86f93dd40a510e8bd6e59a9e9fb95 |                     |
   |                              | ok       | missingness-excludes-entire              |                                  | .lmiss.Variant-exclude-list.SNPQCI.3.txt       | 8585b02980800bb6bb3a8c568df5eb02 |                     |
   |                              | ok       | missingness_entire.lmiss                 |                                  | DKTNF_POPGEN_GS_unQCed.entire_collection.lmiss | 8015ca43097fc6f1fad1d7043e5e758a |                     |
   | generate_hwe_diag            | ok       | DKTNF_POPGEN_GS_hardy.hwe                | dbb09276363f076e388e0ce50f53ac35 | DKTNF_POPGEN_GS_unQCed_hardy.hwe               | dbb09276363f076e388e0ce50f53ac35 | Falscher Vergleich  |
   |                              | ok       | Bilder                                   |                                  |                                                |                                  |                     |
   | exclude_lists_failed_hwe     | ok       | Bilder                                   |                                  |                                                |                                  |                     |
   |                              | ok       | DKTNF_POPGEN_GS_exclude-per-batch        |                                  | SNPQCI.2.failed2plusbatches.txt                | 019d8e70af663a2ae4bb13493791f12b |                     |
   |                              | ok       | DKTNF_POPGEN_GS_exclude-whole-collection |                                  | SNPQCI.1.worstbatchremoved.txt                 | 6d62a4131666bb3f788c8560098ce76f |                     |
   | exclude_bad_variants         | !        | variant-excludes                         |                                  |                                                |                                  | ?                   |
   |                              | ok       | DKTNF_POPGEN_GS_QCI.bed                  |                                  | DKTNF_POPGEN_GS_SNPQCI.bed                     | f3d0ee2f160331832ae09bc2414b5bd5 |                     |
   | 7a/7fa454 draw_def_after_QCI | ok       | DKTNF_POPGEN_GS_entire_collection.hwe    |                                  | DKTNF_POPGEN_GS_SNPQCI_hardy.hwe               | a8f0edcc649b907c02720e9c620f782f |                     |

** SampleQCI.nf   
   | Process              | Verified | Main result file                                       | Hash | V. Result file                                       | V. Hash                          | Problem             |
   |----------------------+----------+--------------------------------------------------------+------+------------------------------------------------------+----------------------------------+---------------------|
   | apply_precalc        | ok       | manually-removed.bed                                   |      | SampleQCI/DKTNF_POPGEN_GS_SNPQCI                     | f3d0ee2f160331832ae09bc2414b5bd5 |                     |
   | determine_miss_het   | ok       | DKTNF_POPGEN_GS_SampleQCI_het.het                      |      | DKTNF_POPGEN_GS_SNPQCI_het.het                       | c5f41ad004fbddc3928f99b83ac8a896 | 2b/55fb5f           |
   |                      | ok       | DKTNF_POPGEN_GS_SampleQCI_miss.imiss                   |      | DKTNF_POPGEN_GS_SNPQCI_miss.imiss                    | d3a57f52d4759d811455407b8a28de4d | Plink-Version       |
   |                      | ok       | DKTNF_POPGEN_GS_SampleQCI_miss.lmiss                   |      | DKTNF_POPGEN_GS_SNPQCI_miss.lmiss                    | e0f396ccbc792a08796f5f16dbde558a |                     |
   |                      | ok       | miss_outliers.txt                                      |      | SNPQCI_miss.outlier                                  | 3cf9f70c440df82480971da1f6a7f9dd | Perl Regex Escaping |
   |                      | ok       | het_outliers.txt                                       |      | SNPQCI_het.het.outlier.txt                           | 24e3b8e2cc0d1fa7632ad7e7f61e1ea3 |                     |
   | calc_pi_hat          | ok       | DKTNF_POPGEN_GS_SampleQCI_pruned.bed                   |      | DKTNF_POPGEN_GS_SNPQCI_pruned.bed                    | 883c7b6029ea43e7272b06aa6ccdab17 |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SNPQCI.bed (input)                     |      |                                                      |                                  |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SampleQCI_miss.outlier.txt (input)     |      |                                                      |                                  |                     |
   | merge_with_hapmap    | ok       | include_variants                                       |      | SNPQCI_chr1-22noLDRegionsnoATandGCnoIndels.txt       | d4b462b1130380232ebaef86d6a03ec4 | 85/b55e5e           |
   |                      | ok       | DKTNF_POPGEN_GS_SampleQCI_pruned_hapmap.bed            |      | DKTNF_POPGEN_GS_SampleQCI_pruned_hapmap.bed          | 63a7e68532192045f78deb7d05c758cc |                     |
   | pca_convert          | ok       | DKTNF_POPGEN_GS_SampleQCI_pruned.eigenstratgeno        |      | DKTNF_POPGEN_GS_SampleQCI_pruned.eigenstratgeno      | ac9feac680ed0485f578795513cac382 |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SNPQCI_pruned_hapmap.eigenstratgeno    |      | DKTNF_POPGEN_GS_SNPQCI_pruned_hapmap2.eigenstratgeno | 6ed444c6c8cce7eb930f595ed89de95f |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SNPQCI_pruned.ind                      |      | DKTNF_POPGEN_GS_SNPQCI_pruned.ind                    | fbcd6eb9ed3298c77488dbd046645106 |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SNPQCI_pruned_hapmap.ind               |      | DKTNF_POPGEN_GS_SNPQCI_pruned_hapmap2.ind            | b8b857debb1bdcd7742d884dd39e10cb |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SNPQCI_pruned.snp                      |      | DKTNF_POPGEN_GS_SNPQCI_pruned.snp                    | identisch bis auf whitespace     |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SNPQCI_pruned_hapmap.snp               |      | DKTNF_POPGEN_GS_SNPQCI_pruned_hapmap2.snp            | identisch bis auf whitespace     |                     |
   | calc_imiss           | ok       | DKTNF_POPGEN_GS_SampleQCI_miss.imiss                   |      | DKTNF_POPGEN_GS_SNPQCI_pruned_miss.imiss             | 776203bd005d793edee9369f3b59ff66 |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SampleQCI_miss.lmiss                   |      | DKTNF_POPGEN_GS_SNPQCI_pruned_miss.lmiss             | f70b91ee24f66dda44f264326a7c6520 |                     |
   | flashpca2_pruned     | ok       | DKTNF_POPGEN_GS_SampleQCI_pruned_1kG_10PC.eval         |      | DKTNF_POPGEN_GS_SNPQCI_pruned_10PC.eval              | 039d0ab9a2ed003d8352a2e8b0a11e57 |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SampleQCI_pruned_10PC.pca.evec         |      | DKTNF_POPGEN_GS_SNPQCI_pruned_10PC.pca.evec          | Nur Rundungsfehler               |                     |
   |                      | ok       | DKTNF_POPGEN_GS_SampleQCI_pruned_10PC.country.pca.evec |      | DKTNF_POPGEN_GS_SNPQCI_pruned_10PC.country.pca.evec  | Nur Rundungsfehler               |                     |
   | flashpca2_pruned_1kg | ok       | *_flashpca2                                            |      |                                                      | Nur Rundungsfehler               |                     |
   |                      |          |                                                        |      |                                                      |                                  |                     |



   | pca_convert               |          | *.ind *.snp                                            |      |                                                      | identisch bis auf whitespace                  |                     |
   |                           |          |                                                        |      |                                                      |                                               |                     |
   | pca_run                   | ok       |                                                        |      |                                                      | alle gleich bis auf log uns bilder            |                     |
   | second_pca_eigenstrat     | ok       | N/A                                                    |      |                                                      |                                               |                     |
   | ibs_merge_and_verify      | ok?      |                                                        |      |                                                      | mein IBS hat eine Zeile mehr, sonst gleich    |                     |
   | detect_duplicates_related | ok       | DKTNF_POPGEN_GS_SampleQCI_pruned_duplicates.txt        |      | DKTNF_POPGEN_GS_SNPQCI_pruned_remove.duplicates.txt  | c9abc616052cd7d481c1fa88e5ca379a              |                     |
   |                           | ok       | DKTNF_POPGEN_GS_SampleQCI_pruned_flag_relatives.txt    |      | DKTNF_POPGEN_GS_SNPQCI_pruned_flag.relatives.txt     | 6eeca813d6734518141d8dcd5f50c67e              |                     |
   |                           | ok       | DKTNF_POPGEN_GS_SampleQCI_pruned_IBS.genome.Z0.Z1      |      | DKTNF_POPGEN_GS_SNPQCI_pruned_IBS.Z0.Z1.genome       | mein genome hat eine Zeile mehr, sonst gleich |                     |
   |                           |          |                                                        |      |                                                      |                                               |                     |

